File: app/static/forecast.js
/* Forecast: deterministic rules over verified data plus an optional external model.
   The API key never leaves the browser: calls go directly to the model provider,
   and the model itself may fetch public site data through declared tools. */
(()=>{
 const state={data:null,loading:false,request:0,abort:null};
 const KEY_STORAGE='gonka-forecast-llm-key';
 const MODEL_STORAGE='gonka-forecast-llm-model';
 const OPENBROKER='https://api.openbroker.gonka.gg/v1/chat/completions';
 const MODEL='deepseek-ai/DeepSeek-V4-Flash-0731';
 const MAX_TOKENS=33000;
 const MAX_TOOL_ROUNDS=6;
 const direction={up:'▲ рост',down:'▼ падение',flat:'— нейтрально'};
 const usd=raw=>(Number(raw)/1e6).toLocaleString('ru-RU',{maximumFractionDigits:0});
 const wgnk=raw=>(Number(raw)/1e9).toLocaleString('ru-RU',{maximumFractionDigits:0});
 function renderVerdict(data){
  const verdict=data.verdict;
  if(!verdict){el('forecast-verdict').innerHTML='<p class="empty">Снимок прогноза ещё не готов: ждём верифицированную историю и порох.</p>';return;}
  el('forecast-verdict').innerHTML='<strong class="'+verdict.direction+'">'+esc(direction[verdict.direction])+
   '</strong><span class="forecast-score">балл правил '+verdict.score.toFixed(3)+
   ' · от −1 до +1 · снимок #'+count(data.snapshot?.height||0)+' · '+clock(data.snapshot?.ts||data.now)+' '+date(data.snapshot?.ts||data.now)+'</span>';
 }
 function renderRules(data){
  el('forecast-rules').innerHTML=(data.rules||[]).map(rule=>
   '<div class="rule"><div><strong>'+esc(rule.name)+'</strong><small>'+esc(rule.explanation)+'</small></div>'+
   '<span class="rule-direction '+rule.direction+'">'+esc(direction[rule.direction].replace('▲ ','').replace('▼ ','').replace('— ',''))+'</span>'+
   '<span class="rule-weight">вес '+rule.weight+'</span></div>').join('')||'<p class="empty">Правила ещё не посчитаны.</p>';
 }
 function renderPowder(data){
  const powder=data.powder;
  if(!powder?.ready){el('powder-buy').innerHTML=el('powder-sell').innerHTML='<p class="empty">Порох ещё собирается…</p>';return;}
  const main=n=>n>1?' <small title="Вероятная группа одного участника; показан основной адрес">· основной из '+n+'</small>':'';
  const buyers=(powder.buyers||[]).slice(0,8).map(row=>
   '<li><button class="address-button" data-flow-address="'+esc(row.address)+'" title="'+esc(row.address)+'">'+esc(short(row.address,8))+'</button>'+
   '<span><strong>'+usd(row.own_raw)+'</strong> USDT'+(Number(row.chain_raw)>0?' <small>+ цепочки '+usd(row.chain_raw)+'</small>':'')+main(row.group_size||1)+'</span></li>').join('');
  const sellers=(powder.sellers||[]).slice(0,8).map(row=>
   '<li><button class="address-button" data-flow-address="'+esc(row.address)+'" title="'+esc(row.address)+'">'+esc(short(row.address,8))+'</button>'+
   '<span><strong>'+wgnk(row.wgnk_raw)+'</strong> WGNK'+(Number(row.gnk_raw)>0?' <small>+ GNK '+wgnk(row.gnk_raw)+'</small>':'')+main(row.group_size||1)+'</span></li>').join('');
  const t=powder.totals;
  el('powder-buy').innerHTML='<h3>Порох покупателей</h3><div class="total">'+usd(t.buy_own_raw)+
   ' <span class="unit">USDT-экв.</span></div><p class="unit">собственный + '+usd(t.buy_chain_raw)+' в цепочках финансирования</p>'+
   '<ul class="powder-list">'+(buyers||'<li><small>Нет адресов над порогом.</small></li>')+'</ul>';
  el('powder-sell').innerHTML='<h3>Запасы продавцов</h3><div class="total">'+wgnk(t.sell_wgnk_raw)+
   ' <span class="unit">WGNK</span></div><p class="unit">+ '+wgnk(t.sell_gnk_raw)+' GNK у мост-источников · эскроу '+wgnk(t.escrow_raw)+' GNK</p>'+
   '<ul class="powder-list">'+(sellers||'<li><small>Нет адресов над порогом.</small></li>')+'</ul>';
 }

 // ---- Tools the model may call; every fetch is same-origin public API data ----
 const TOOLS=[
  {type:'function',function:{name:'forecast_snapshot',description:'Полный снимок прогноза: пять правил с числами, сухой порох покупателей (адреса, собственные и цепочечные стейблкоины, покрытие балансов), запасы продавцов WGNK/GNK.',
   parameters:{type:'object',properties:{},required:[]}}},
  {type:'function',function:{name:'daily_history',description:'Дневная история: средневзвешенная цена USDT, объёмы покупок и продаж WGNK и глубина ликвидности ±2% пула 30 б.п. по дням.',
   parameters:{type:'object',properties:{days:{type:'integer',description:'Сколько последних дней вернуть, до 120',minimum:1,maximum:120}},required:[]}}},
  {type:'function',function:{name:'trader_leaders',description:'Крупнейшие покупатели или продавцы WGNK за всю историю, с группами вероятных адресов одного участника.',
   parameters:{type:'object',properties:{side:{type:'string',enum:['buy','sell']},limit:{type:'integer',minimum:1,maximum:50}},required:['side']}}},
  {type:'function',function:{name:'bridge_feed',description:'Последние операции моста: выпуск WGNK в Ethereum (минты) и вывод в Gonka (бёрны) с суммами и временем.',
   parameters:{type:'object',properties:{limit:{type:'integer',minimum:1,maximum:100}},required:[]}}}];

 async function getJSON(url){
  let response=null,text='';
  for(let attempt=0;attempt<3;attempt++){
   if(attempt)await new Promise(done=>setTimeout(done,attempt*3000));
   try{
    response=await fetch(url);
    text=await response.text();
    if(response.ok&&!text.trim().startsWith('<'))return JSON.parse(text);
   }catch(error){response=null;}
  }
  if(response&&!response.ok&&response.status>=500)
   throw new Error('сайт перезапускается (HTTP '+response.status+') — подождите минуту и повторите');
  throw new Error('данные сайта не ответили — подождите минуту и повторите');
 }
 async function runTool(call){
  const args={};
  try{Object.assign(args,JSON.parse(call.function.arguments||'{}'));}catch{}
  switch(call.function.name){
   case 'forecast_snapshot':{
    const f=await getJSON('/api/mints/forecast');
    const powder=f.powder||{};
    return JSON.stringify({price_usdt:f.price_raw?Number(f.price_raw)/1e12:null,
     verdict:f.verdict,rules:(f.rules||[]).map(r=>({name:r.name,score:r.score,weight:r.weight,explanation:r.explanation})),
     powder:{coverage:powder.balance_coverage,totals:powder.totals,
      top_buyers:(powder.buyers||[]).slice(0,10).map(b=>({address:b.address,own_usdt:+(Number(b.own_raw)/1e6).toFixed(0),chain_usdt:+(Number(b.chain_raw)/1e6).toFixed(0)})),
      top_sellers:(powder.sellers||[]).slice(0,10).map(s=>({address:s.address,wgnk:+(Number(s.wgnk_raw)/1e9).toFixed(0),gnk:+(Number(s.gnk_raw)/1e9).toFixed(0)}))}});
   }
   case 'daily_history':{
    const days=Math.min(120,Math.max(1,args.days||30));
    const f=await getJSON('/api/mints/flows?side=all&limit=1');
    const pool30=(f.pools||[]).find(p=>p.fee===3000);
    const bands=f.liquidity_bands&&pool30?f.liquidity_bands[pool30.address]:null;
    const band=(bands&&bands['2'])||[];
    const byDate=new Map(band.map(b=>[b.date,b]));
    return JSON.stringify((f.daily||[]).slice(-days).map(d=>{
     const b=byDate.get(d.date)||{};
     return {date:d.date,price_usdt:+(Number(d.price_raw)/1e12).toFixed(6),
      bought_wgnk:+(Number(d.bought_raw)/1e9).toFixed(0),sold_wgnk:+(Number(d.sold_raw)/1e9).toFixed(0),
      buy_usdt:+(Number(d.buy_quote_raw)/1e6).toFixed(0),sell_usdt:+(Number(d.sale_quote_raw)/1e6).toFixed(0),
      band_wgnk:b.wgnk_raw?+(Number(b.wgnk_raw)/1e9).toFixed(0):null,
      band_usdt:b.usdt_raw?+(Number(b.usdt_raw)/1e6).toFixed(0):null};}));
   }
   case 'trader_leaders':{
    const side=args.side==='sell'?'sell':'buy';
    const limit=Math.min(50,Math.max(1,args.limit||15));
    const l=await getJSON('/api/mints/leaders?grouped=true');
    return JSON.stringify((l[side==='buy'?'buyers':'sellers']||[]).slice(0,limit).map(r=>({
     address:r.group?r.group.addresses:r.address,volume_wgnk:+(Number(r.volume_raw)/1e9).toFixed(0),
     quote_usdt:+(Number(r.quote_raw)/1e6).toFixed(0),swaps:r.swaps,
     group:r.group?{size:r.group.addresses.length,evidence:r.group.evidence.map(e=>e.label)}:null})));
   }
   case 'bridge_feed':{
    const limit=Math.min(100,Math.max(1,args.limit||30));
    const b=await getJSON('/api/mints/bridge?minimum=1000&limit='+limit);
    return JSON.stringify((b.items||[]).map(e=>({kind:e.kind==='bridge_mint'?'mint(WGNK в Ethereum)':'burn(вывод в Gonka)',
     address:e.address,amount_wgnk:+(Number(e.amount_raw)/1e9).toFixed(0),
     time:new Date(e.ts*1000).toISOString().slice(0,16).replace('T',' ')})));
   }
   default:
    return JSON.stringify({error:'unknown tool '+call.function.name});
  }
 }

 const SYSTEM_PROMPT=[
  'Ты исследователь рынка токена WGNK (Gonka). Твоя задача — саммари рынка в стиле независимого исследовательского отчёта, а не прогноз.',
  'Сначала собери данные инструментами (обычно 2-4 вызова: снимок прогноза, дневная история, рейтинги, лента моста), затем напиши на русском связный отчёт со структурой:',
  '1. Вводный абзац: дата и время снимка, текущая цена и изменение за 24 часа, итоговая оценка состояния рынка одним предложением с выделенным выводом (жирным **...**).',
  '2. «Что произошло»: объёмы покупок и продаж WGNK/USDT за 24 часа и 7 дней, перевес покупок, движение цены.',
  '3. «Покупатели»: крупнейшие покупатели из рейтинга с точными суммами WGNK и USDT, упомяни группы вероятных одного участника (адреса перечисляй сокращённо, например 0x1234…abcd), порох покупателей (стейблкоины собственные и в цепочках) и степень покрытия балансов.',
  '4. «Продавцы и мост»: крупнейшие продавцы с их запасами WGNK/GNK, операции моста за 24 часа (минты против бёрнов) и что это значит для предложения.',
  '5. «Ликвидность»: глубина книги в диапазонах ±2% / ±5% / ±10% / ±20% на последний день и как она менялась.',
  '6. «Оценка»: 3-5 пунктов выводов с оговорками о данных (концентрация, неполное покрытие пороха, связи по переводам — вероятность, не владелец).',
  'Стиль: плотный фактический, каждое утверждение с числом из полученных данных, никаких выдуманных чисел и адресов, никаких советов покупать или продавать, без дисклеймеров про не-финансовый-совет (это саммари данных). Объём 350-700 слов.'].join(' ');

 const llm={busy:false};
 function currentModel(){
  const select=el('forecast-llm-model');
  return select&&select.value?select.value:MODEL;
 }
 async function postModel(key,body){
  // "Failed to fetch" is a dropped connection before any response: worth retrying.
  let lastError=null;
  for(const wait of [0,3000,8000]){
   if(wait)await new Promise(done=>setTimeout(done,wait));
   try{return await fetch(OPENBROKER,{method:'POST',
    headers:{'Content-Type':'application/json','Authorization':'Bearer '+key},
    body:JSON.stringify(body)});}
   catch(error){lastError=error;}
  }
  throw new Error('нет связи с api.openbroker.gonka.gg — проверьте интернет и повторите запрос');
 }
 async function callModel(key,messages,tools){
  const base={model:currentModel(),temperature:0,max_tokens:MAX_TOKENS,messages,
   tools:tools||undefined,tool_choice:tools?'auto':undefined};
  // Мышление high у всех агентов, провайдер один (OpenBroker): неизвестное
  // поле даёт 400 — пробуем известные формы параметра, затем обычный запрос.
  const attempts=[{...base,reasoning:{effort:'high'}},{...base,reasoning_effort:'high'},base];
  let lastError=null;
  for(const body of attempts){
   const response=await postModel(key,body);
   if(response.ok){
    let payload;const text=await response.text();
    try{payload=JSON.parse(text);}catch{throw new Error('модель ответила не JSON');}
    const choice=payload.choices?.[0];
    const content=(choice?.message?.content||'').trim();
    if(!content&&choice?.finish_reason==='length')
     throw new Error('модель исчерпала лимит вывода на размышление — повторите запрос');
    return {content,tool_calls:choice?.message?.tool_calls||null};
   }
   let detail='HTTP '+response.status;let concurrent=false;
   try{const payload=await response.json();
    const message=payload.error?.message||String(payload.error||payload.detail||'');
    detail+=': '+message;concurrent=/concurrent/i.test(message);}catch{}
   if(response.status===400){lastError=new Error(detail);continue;}
   if(response.status===429&&!concurrent)detail+=' · лимит запросов ключа OpenBroker: подождите минуту и повторите, либо проверьте квоту/тариф ключа';
   if(response.status===401)detail='Ключ не принят (401): проверьте, что скопировали ключ OpenBroker целиком';
   const error=new Error(detail);error.concurrent=concurrent;error.status=response.status;throw error;
  }
  throw lastError||new Error('модель не ответила');
 }
 async function askModel(){
  if(llm.busy){el('forecast-llm-status').textContent='Запрос уже выполняется…';return;}
  const key=el('forecast-llm-key').value.trim();
  const status=el('forecast-llm-status');
  if(!key){status.textContent='Введите ключ API.';return;}
  try{localStorage.setItem(KEY_STORAGE,key);}catch{}
  const data=state.data;
  if(!data?.ready){status.textContent='Снимок прогноза ещё не готов.';return;}
  llm.busy=true;el('forecast-llm-ask').disabled=true;
  try{
   const messages=[{role:'system',content:SYSTEM_PROMPT},
    {role:'user',content:'Составь саммари рынка WGNK на сейчас. Вердикт правил сайта: '+
      (data.verdict?.label||'?')+' ('+(data.verdict?.score??0).toFixed(3)+'). Собери данные инструментами и напиши отчёт по заданной структуре.'}];
   let content='',rounds=0;const toolNotes=[];
   let answer;
   while(rounds++<MAX_TOOL_ROUNDS){
    answer=await callModel(key,messages,TOOLS);
    if(answer.tool_calls&&answer.tool_calls.length){
     messages.push({role:'assistant',content:answer.content||'',tool_calls:answer.tool_calls});
     for(const call of answer.tool_calls){
      status.textContent='Агент запрашивает: '+call.function.name+'…';
      const result=await runTool(call);
      messages.push({role:'tool',tool_call_id:call.id,content:String(result).slice(0,60000)});
      toolNotes.push(call.function.name);
     }
     continue;
    }
    content=answer.content;
    break;
   }
   if(!content){
    // Provider may not support tool calls: one rich request without tools.
    status.textContent='Инструменты не поддержаны, спрашиваем одним запросом…';
    const f=await getJSON('/api/mints/forecast');
    const flows=await getJSON('/api/mints/flows?side=all&limit=1');
    const rich=JSON.stringify({verdict:f.verdict,rules:(f.rules||[]).map(r=>[r.name,r.score,r.explanation]),
     powder:f.powder&&{totals:f.powder.totals,coverage:f.powder.balance_coverage,
      buyers:(f.powder.buyers||[]).slice(0,10),sellers:(f.powder.sellers||[]).slice(0,10)},
     daily_30d:(flows.daily||[]).slice(-30).map(d=>[d.date,+(Number(d.price_raw)/1e12).toFixed(6),+(Number(d.bought_raw)/1e9).toFixed(0),+(Number(d.sold_raw)/1e9).toFixed(0)])}).slice(0,50000);
    content=(await callModel(key,[{role:'system',content:SYSTEM_PROMPT},{role:'user',content:rich}],null)).content;
    toolNotes.push('без инструментов');
   }
   el('forecast-llm-reply').hidden=false;
   el('forecast-llm-reply').innerHTML=formatSummary(content||'(пустой ответ)');
   status.textContent='Готово · прямой вызов из браузера · модель '+currentModel()+
    (toolNotes.length?' · данные: '+toolNotes.join(', '):'');
  }finally{llm.busy=false;el('forecast-llm-ask').disabled=false;}
 }
 function formatSummary(text){
  // Minimal safe markdown: bold and paragraphs; everything else stays literal.
  const html=esc(text).replace(/\*\*([^*]+)\*\*/g,'<strong>$1</strong>');
  const parts=html.split('\n\n');
  return parts.map(p=>'<p>'+p.split('\n').join('<br>')+'</p>').join('');
 }
 el('forecast-llm-ask').addEventListener('click',()=>askModel().catch(error=>{el('forecast-llm-status').textContent='Ошибка: '+error.message;}));
 el('forecast-llm-key').addEventListener('keydown',event=>{if(event.key==='Enter')el('forecast-llm-ask').click();});
 try{el('forecast-llm-key').value=localStorage.getItem(KEY_STORAGE)||'';}catch{}
 try{el('forecast-llm-model').value=localStorage.getItem(MODEL_STORAGE)||MODEL;}catch{}
 el('forecast-llm-model').addEventListener('change',event=>{
  try{localStorage.setItem(MODEL_STORAGE,event.target.value);}catch{}});
 window.addEventListener('gonka:view',()=>{if(!el('forecast-view').hidden)refresh();});
 setInterval(()=>{if(!document.hidden&&!el('forecast-view').hidden)refresh();},60000);
 document.addEventListener('visibilitychange',()=>{if(!document.hidden&&!el('forecast-view').hidden)refresh();});
 async function refresh(){
  if(state.loading)return;
  state.abort?.abort();
  const id=++state.request,controller=state.abort=new AbortController(),timer=setTimeout(()=>controller.abort(),30000);
  state.loading=true;el('forecast-view').setAttribute('aria-busy','true');
  try{
   const response=await fetch('/api/mints/forecast',{signal:controller.signal});
   if(!response.ok)throw new Error('HTTP '+response.status);
   const data=await response.json();if(id!==state.request)return;
   state.data=data;
   el('forecast-status').textContent=data.ready?'':'Снимок прогноза ещё собирается; покажем правила, как только история и порох будут проверены.';
   renderVerdict(data);renderRules(data);renderPowder(data);
  }catch(error){if(id===state.request)el('forecast-status').textContent='Не удалось обновить прогноз. Повторим автоматически.';}
  finally{clearTimeout(timer);if(id===state.request){state.loading=false;el('forecast-view').setAttribute('aria-busy','false');}}
 }
 if(!el('forecast-view').hidden)refresh();
})();

File: app/forecast.py
"""Deterministic up/down rules from verified data plus the dry-powder picture.

Every rule, weight and raw value is returned so the UI can show the reasoning.
This is arithmetic over public chain data, not financial advice.
"""
import time
from collections import defaultdict

from .db import tokens
from .powder import aggregate

OPENBROKER = "https://api.openbroker.gonka.gg/v1/chat/completions"
MODEL = "deepseek-ai/DeepSeek-V4-Flash-0731"

WEIGHTS = {"powder": .30, "flow": .30, "bridge": .15, "reload": .10, "momentum": .15}


def _price(db):
    row = db.conn.execute(
        "SELECT quote_raw,amount_raw FROM events WHERE chain='ethereum' AND finalized=1 "
        "AND kind IN ('buy','sell') ORDER BY ts DESC,height DESC,idx DESC LIMIT 1").fetchone()
    if not row or int(row["amount_raw"]) == 0:
        return None
    return int(row["quote_raw"]) * 10**21 // (int(row["amount_raw"]) * 10**6)


def _flow_windows(db, now):
    result = {}
    for hours in (24, 168):
        since = now - hours * 3600
        rows = db.conn.execute(
            "SELECT kind, SUM(CAST(quote_raw AS INTEGER)) q FROM events "
            "WHERE chain='ethereum' AND finalized=1 AND kind IN ('buy','sell') AND ts>=? GROUP BY kind",
            (since,)).fetchall()
        buy = sum(r["q"] for r in rows if r["kind"] == "buy")
        sell = sum(r["q"] for r in rows if r["kind"] == "sell")
        result[hours] = (buy, sell)
    return result


def _bridge_windows(db, now):
    result = {}
    for hours in (24, 168):
        since = now - hours * 3600
        rows = db.conn.execute(
            "SELECT kind, SUM(CAST(amount_raw AS INTEGER)) v FROM events "
            "WHERE chain='ethereum' AND finalized=1 AND kind IN ('bridge_mint','bridge_burn') AND ts>=? GROUP BY kind",
            (since,)).fetchall()
        minted = sum(r["v"] for r in rows if r["kind"] == "bridge_mint")
        burned = sum(r["v"] for r in rows if r["kind"] == "bridge_burn")
        result[hours] = (minted, burned)
    return result


def _reload_delta(db, now):
    rows = db.conn.execute(
        "SELECT ts, buy_own_raw, buy_chain_raw FROM powder_snapshots ORDER BY ts DESC LIMIT 120").fetchall()
    if not rows:
        return None
    latest = int(rows[0]["buy_own_raw"]) + int(rows[0]["buy_chain_raw"])
    for row in rows:
        if now - row["ts"] >= 24 * 3600:
            return latest - (int(row["buy_own_raw"]) + int(row["buy_chain_raw"]))
    oldest = rows[-1]
    if rows[0]["ts"] - oldest["ts"] >= 12 * 3600:
        return latest - (int(oldest["buy_own_raw"]) + int(oldest["buy_chain_raw"]))
    return None


def _momentum(db, now):
    row = db.conn.execute(
        "SELECT ts, quote_raw, amount_raw FROM events WHERE chain='ethereum' AND finalized=1 "
        "AND kind IN ('buy','sell') AND ts<=? ORDER BY ts DESC LIMIT 1", (now - 24 * 3600,)).fetchone()
    price = _price(db)
    if not row or price is None or int(row["amount_raw"]) == 0:
        return None
    then = int(row["quote_raw"]) * 10**21 // (int(row["amount_raw"]) * 10**6)
    if then == 0:
        return None
    return (price - then) * 10000 // then  # basis points


def snapshot(db):
    """The full rule set and verdict at the current verified state."""
    from .flows import market_packet
    now = int(time.time())
    packet = market_packet(db)
    snap = packet["snapshot"] if packet else db.get("flow:snapshot")
    result = {"ready": False, "now": now, "snapshot": snap, "rules": [], "verdict": None,
              "weights": WEIGHTS, "price_raw": None, "powder": None,
              "disclaimer": "Арифметика по публичным данным блокчейна. Не финансовый совет."}
    if not snap or not snap.get("ledger_verified"):
        return result
    price = _price(db)
    result["price_raw"] = str(price) if price is not None else None
    powder = aggregate(db)
    result["powder"] = powder
    if price is None:
        return result
    rules = []

    def add(key, name, score, explanation):
        rules.append({"key": key, "name": name, "weight": WEIGHTS[key],
                      "score": round(score, 3),
                      "direction": "up" if score > 0 else "down" if score < 0 else "flat",
                      "explanation": explanation})

    # 1. Dry powder versus seller overhang at the current price.
    if not powder.get("ready"):
        add("powder", "Порох против запасов", 0, "Порох ещё не собран")
    else:
        totals = powder["totals"]
        coverage = powder.get("balance_coverage") or {}
        covered, qualified = coverage.get("covered", 0), coverage.get("qualified", 0)
        powder_raw = int(totals["buy_own_raw"]) + int(totals["buy_chain_raw"])
        overhang = int(totals["sell_wgnk_raw"]) + int(totals["escrow_raw"]) + int(totals["sell_gnk_raw"])
        overhang_raw = overhang * price // 10**15  # 1e9-raw GNK x 1e12-raw price -> 1e6-raw USDT
        if qualified and (covered < 3 or covered * 2 < qualified):
            add("powder", "Порох против запасов", 0,
                f"Балансы держателей обновляются ({covered} из {qualified}) — правило пока не голосует")
        elif not overhang_raw:
            add("powder", "Порох против запасов", 0,
                f"Порох {tokens(powder_raw, 6)} USDT-экв; запасы продавцов не оценены")
        else:
            ratio = powder_raw / overhang_raw
            score = 1 if ratio >= 2 else .5 if ratio >= 1 else -.5 if ratio >= .5 else -1
            add("powder", "Порох против запасов", score,
                f"Порох {tokens(powder_raw, 6)} USDT-экв против запасов ≈{tokens(overhang_raw, 6)} "
                f"USDT-экв (отношение {ratio:.2f})")

    # 2. Net buy pressure in verified swaps.
    flows = _flow_windows(db, now)
    buy, sell = flows[24]
    total = buy + sell
    share = (buy - sell) / total if total else 0
    weekly = flows[168]
    weekly_share = (weekly[0] - weekly[1]) / (weekly[0] + weekly[1]) if weekly[0] + weekly[1] else 0
    add("flow", "Давление покупок 24ч/7д", max(-1, min(1, (share + weekly_share) / .6)),
        f"24ч: покупки {tokens(buy, 6)} против продаж {tokens(sell, 6)} (перевес {share:+.0%}); "
        f"7д перевес {weekly_share:+.0%}")

    # 3. Bridge direction: mints add sell supply, burns are accumulation.
    mints = _bridge_windows(db, now)
    minted, burned = mints[24]
    if minted or burned:
        share = (burned - minted) / (minted + burned)
        add("bridge", "Направление моста 24ч", max(-1, min(1, share / .5)),
            f"Выпущено {tokens(minted)} WGNK, выведено {tokens(burned)} WGNK за 24ч")
    else:
        add("bridge", "Направление моста 24ч", 0, "За 24ч мост не двигался")

    # 4. Dry powder change over the last day.
    delta = _reload_delta(db, now)
    if delta is None:
        add("reload", "Изменение пороха за 24ч", 0, "Снимков пока недостаточно")
    else:
        add("reload", "Изменение пороха за 24ч", .5 if delta > 0 else -.5 if delta < 0 else 0,
            f"Порох покупателей изменился на {tokens(delta, 6)} USDT-экв")

    # 5. Price momentum, capped at ±20%.
    change = _momentum(db, now)
    if change is None:
        add("momentum", "Цена за 24ч", 0, "Нет сделки сутки назад")
    else:
        add("momentum", "Цена за 24ч", max(-1, min(1, change / 2000)),
            f"{change / 100:+.2f}% за 24 часа")

    score = sum(rule["score"] * rule["weight"] for rule in rules)
    verdict = "up" if score >= .15 else "down" if score <= -.15 else "flat"
    labels = {"up": "рост", "down": "падение", "flat": "нейтрально"}
    result.update(ready=True, rules=rules,
                  verdict={"direction": verdict, "label": labels[verdict],
                           "score": round(score, 3)})
    return result


def model_prompt(data):
    """Compact numeric context for an optional external model call."""
    rules = "; ".join(f"{r['name']}={r['score']:+.2f}" for r in data.get("rules") or [])
    verdict = (data.get("verdict") or {})
    powder = (data.get("powder") or {})
    totals = powder.get("totals") or {}
    price = tokens(int(data["price_raw"]), 12) if data.get("price_raw") else "?"
    return (
        "Ты аналитик рынка токена WGNK (Gonka). По нижеприведённым проверенным данным "
        "ответь одной строкой: ВЕРДИКТ: рост или падение или нейтрально — и одно предложение "
        "объяснения на русском.\n"
        f"Цена сейчас: {price} USDT. Правила (оценка от -1 до 1): {rules}. "
        f"Итог правил: {verdict.get('label', '?')} ({verdict.get('score', 0):+.2f}). "
        f"Порох покупателей: собственный {totals.get('buy_own_raw', '?')} и в цепочках "
        f"{totals.get('buy_chain_raw', '?')} (сырые единицы 1e-6). Запасы продавцов WGNK "
        f"{totals.get('sell_wgnk_raw', '?')} (1e-9), GNK {totals.get('sell_gnk_raw', '?')} "
        f"(1e-9), эскроу {totals.get('escrow_raw', '?')} (1e-9).")
